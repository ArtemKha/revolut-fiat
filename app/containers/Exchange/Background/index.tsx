import React from 'react';
import { Reset, FullScreen, Centered, Text, Gradient } from './styled';

export default function Background({ children }): React.ComponentProps {
  const [hue, setHue] = React.useState(240);
  const [accelerator, setAccelerator] = React.useState(1);

  setTimeout(() => {
    if (hue > 240) {
      setAccelerator(-1);
    }
    if (hue < 180) {
      setAccelerator(1);
    }
    setHue(hue + accelerator);
  }, 2350);

  return (
    <>
      <Reset />
      <FullScreen>
        {children}
        <Gradient hue={hue} />
      </FullScreen>
    </>
  );
}
