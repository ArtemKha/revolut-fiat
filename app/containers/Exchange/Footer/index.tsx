import React from 'react';
import { Container } from './styled';

export default function Footer({ children }): React.FunctionComponentElement {
  return <Container>{children}</Container>;
}
