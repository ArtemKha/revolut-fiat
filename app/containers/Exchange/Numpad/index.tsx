import React from 'react';
import { CaretLeftOutline } from '@ant-design/icons';
import { Numbers, Cell, Container, Delete } from './styled';
import { Icon } from 'components/Icon';

interface IAppProps {
  onInput: (v: string) => void;
}
const Numpad: React.FC<IAppProps> = ({ onInput }) => {
  const pushValue = e => {
    const value = e.target.innerText;
    onInput(value);
  };
  return (
    <Container>
      <Numbers onClick={pushValue} data-testid="numpad-input">
        {[1, 2, 3, 4, 5, 6, 7, 8, 9, '.', 0].map(value => (
          <Cell key={value} data-testid={value}>
            {value}
          </Cell>
        ))}
        <Cell data-testid="del">
          <Delete>
            <Icon icon={CaretLeftOutline} />
          </Delete>
        </Cell>
      </Numbers>
    </Container>
  );
};

export default Numpad;
