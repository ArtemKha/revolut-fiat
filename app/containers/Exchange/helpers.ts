export function isValidValue(value, pocket) {
  // definitely needs test, but time is limited ¯\_(ツ)_/¯

  const isValidCurrencyAmount = /^-?\d*[.,]?\d{0,2}$/.test(value);
  const isNull = /^0\d/.test(value);
  const isAmountAvailable = value === '' || pocket.value >= parseFloat(value);

  return isValidCurrencyAmount && !isNull && isAmountAvailable;
}

export const numToString = num => num.toString().replace(',', '.');
