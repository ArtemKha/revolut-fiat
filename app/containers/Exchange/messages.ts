import { defineMessages } from 'react-intl';

export const scope = 'exchange';

export default defineMessages({
  have: {
    id: `${scope}.have`,
    defaultMessage: 'You have',
  },
  exchange: {
    id: `${scope}.exchange`,
    defaultMessage: 'Exchange',
  },
  cancel: {
    id: `${scope}.cancel`,
    defaultMessage: 'Cancel',
  },
});
