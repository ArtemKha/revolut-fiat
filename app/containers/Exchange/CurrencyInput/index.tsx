import React, { useRef, useState } from 'react';
import { Label, Description, InputLine, Input } from './styled';
import messages from 'containers/Exchange/messages';
import { FormattedMessage } from 'react-intl';
import Rate from '../Rate';
import { Pocket } from 'containers/App/types';

interface Props {
  id: string;
  relation: number;
  currencies: Pocket[];
  currency: Pocket;
  isTop?: boolean;
  hasError: boolean;
  setRefToUse: any;
  value: string;
  setValue: (value: string) => void;
}
const CurrencyInput: React.FC<Props> = ({
  id,
  relation,
  setRefToUse,
  currencies,
  currency,
  value,
  hasError,
  setValue,
  isTop = false,
}) => {
  const [unit, fraction] = currency.value.toString().split('.');
  const fractional = fraction ? '.' + fraction : '';

  const key = currency.key.toUpperCase();
  const inputRef = useRef<HTMLInputElement>(null);
  const prefix = value ? (isTop ? '-' : '+') : '';

  function onChange(e) {
    const { value } = e.target;

    // remove + and - signs before validate
    const stringValue = value
      .split('')
      .filter(char => !['-', '+'].includes(char))
      .join('');

    setValue(stringValue);
  }

  function onFocus() {
    setRefToUse(inputRef);
  }

  return (
    <>
      <InputLine>
        <Label htmlFor={id} data-testid={id + '-label'}>
          {key}
        </Label>
        <Input
          id={id}
          className={hasError ? 'shake' : ''}
          ref={inputRef}
          value={prefix + value}
          onChange={onChange}
          onFocus={onFocus}
          type="text"
          maxLength={11}
          size={11}
          disabled={!isTop}
        />
      </InputLine>
      <InputLine>
        <Description>
          <FormattedMessage {...messages.have} /> {unit}
          {fractional} {key}
        </Description>
        <Description>
          {!isTop && (
            <Rate type="incoming" relation={relation} currencies={currencies} />
          )}
        </Description>
      </InputLine>
    </>
  );
};

export default CurrencyInput;
