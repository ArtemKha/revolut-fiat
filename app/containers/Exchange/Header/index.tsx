import React from 'react';
import { CaretLeftOutline } from '@ant-design/icons';
import { Cell, Container, TabText, Low } from './styled';
import { Icon } from 'components/Icon';
import messages from 'containers/Exchange/messages';
import { FormattedMessage } from 'react-intl';
import { withRouter, RouteComponentProps } from 'react-router';
import Rate from 'containers/Exchange/Rate';
import { Pocket } from 'containers/App/types';

interface IAppProps extends RouteComponentProps {
  relation: number;
  currencies: Pocket[];
  onExchange: () => void;
}
const Header: React.FC<IAppProps> = ({
  currencies,
  relation,
  onExchange,
  history,
}) => {
  const toPockets = () => history.push('/');

  return (
    <Container>
      <Cell onClick={toPockets}>
        <FormattedMessage {...messages.cancel} />
      </Cell>
      <Cell>
        <Rate type="outgoing" relation={relation} currencies={currencies} />
      </Cell>
      <Cell onClick={onExchange}>
        <FormattedMessage {...messages.exchange} />
      </Cell>
    </Container>
  );
};

export default withRouter(Header);
