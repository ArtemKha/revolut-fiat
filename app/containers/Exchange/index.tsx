import React, { useEffect, useRef, useState } from 'react';
import { Helmet } from 'react-helmet';
import { FormattedMessage, injectIntl } from 'react-intl';
import { CurrencyExchangeTab, Centered } from './styled';
import Slider from 'components/Slider';
import Input from './CurrencyInput';
import Numpad from './Numpad';
import messages from './messages';
import Header from './Header';
import {
  useSelector as useReduxSelector,
  useDispatch,
  TypedUseSelectorHook,
} from 'react-redux';
import { ApplicationRootState } from 'types';
import { SliderMethods } from './types';
import { setCurrency } from 'containers/App/actions';
import { isValidValue } from './helpers';

const useSelector: TypedUseSelectorHook<
  ApplicationRootState
> = useReduxSelector;

function Exchange({
  intl,
  afterChange = type => {
    return index => {
      return { type, index };
    };
  },
  sliderRefs = [useRef<SliderMethods>(null), useRef<SliderMethods>(null)],
  inputRef = useRef<HTMLInputElement>(null),
}) {
  const dispatch = useDispatch();
  const [pockets, rates, outgoingCurrencyKey] = useSelector(({ global }) => [
    global.pockets,
    global.rates,
    global.currency,
  ]);

  const [outgoingSlider, incomingSlider] = sliderRefs;
  const [outgoingPockets, setOutgoingPockets] = useState(pockets);

  const [outgoingAmount, setOutgoingAmount] = useState('');
  const [incomingAmount, setIncomingAmount] = useState('');

  const [outgoingCurrency, setOutgoingCurrency] = useState(pockets[0]);
  const [incomingCurrency, setIncomingCurrency] = useState(outgoingPockets[0]);

  useEffect(() => {
    // set incoming currency state
    setOutgoingPockets(pockets.filter(o => o.key !== outgoingCurrencyKey));
    const newOutgoingCurrency = pockets.filter(
      o => o.key === outgoingCurrencyKey,
    );
    setOutgoingCurrency(newOutgoingCurrency[0]);

    // reset outgoing currency state
    setOutgoingAmount('');
    setIncomingCurrency(outgoingPockets[0]);
    incomingSlider.current!.slickGoTo(0);
  }, [outgoingCurrencyKey]);

  useEffect(() => {
    // use incomingCurrency here to calculate rate

    const relation = 1;
    setIncomingAmount((Number(outgoingAmount) * relation).toString());
  }, [incomingAmount]);

  function beforeOutgoingChange(_, next) {
    dispatch(setCurrency(pockets[next].key));
  }

  function beforeIncomingChange(_, next) {
    setIncomingCurrency(outgoingPockets[next]);
    console.log(outgoingPockets[next]);
  }

  function onNumpadInput(value) {
    const newValue =
      value === undefined
        ? outgoingAmount.slice(0, -1)
        : outgoingAmount + value;

    console.log(value, newValue);

    if (isValidValue(newValue, outgoingCurrency)) {
      setOutgoingAmount(newValue);
    }
  }

  return (
    <>
      <Helmet>
        <title>{intl.formatMessage(messages.exchange)}</title>
        <meta
          name="description"
          content={intl.formatMessage(messages.exchange)}
        />
      </Helmet>
      <>
        <Centered>
          <Header />
          <CurrencyExchangeTab data-testid="outgoing-slider" isTop={true}>
            <Slider
              refToUse={outgoingSlider}
              beforeChange={beforeOutgoingChange}
              afterChange={afterChange}
            >
              {pockets.map(item => (
                <Input
                  id="outgoing"
                  setRefToUse={ref => (inputRef = ref)}
                  value={outgoingAmount}
                  setValue={setOutgoingAmount}
                  key={item.key}
                  currency={item}
                  isTop={true}
                />
              ))}
            </Slider>
          </CurrencyExchangeTab>
          <CurrencyExchangeTab isTop={false}>
            <Slider
              refToUse={incomingSlider}
              beforeChange={beforeIncomingChange}
              afterChange={afterChange}
            >
              {outgoingPockets.map(item => (
                <Input
                  id="incoming"
                  value={outgoingAmount}
                  setValue={() => null}
                  setRefToUse={ref => null}
                  key={item.key}
                  currency={item}
                  isTop={false}
                />
              ))}
            </Slider>
          </CurrencyExchangeTab>
          <Numpad onInput={onNumpadInput} />
        </Centered>
      </>
    </>
  );
}

export default injectIntl(Exchange);
