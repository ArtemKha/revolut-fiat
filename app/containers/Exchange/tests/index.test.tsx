import React, { RefObject } from 'react';
import { render, sleep } from 'utils/test-utils';
import { cleanup, fireEvent } from '@testing-library/react';

import * as appActions from 'containers/App/actions';
import configureStore from '../../../configureStore';
import Exchange from '../index';
import history from '../../../utils/history';
import { act } from 'react-test-renderer';
import { SliderMethods } from '../types';

jest.mock('containers/App/actions');

const renderExchange = () => render(<Exchange />);

const makeArrangeSet = () => {
  const refs: Array<RefObject<SliderMethods>> = [
    React.createRef(), // outgoing
    React.createRef(), // incoming
  ];
  const afterChange = val => {
    console.log(val);
  };

  return [refs, afterChange];
};

describe('<Exchange />', () => {
  // let store;
  // const mockedLoadRepos = appActions.loadRepos as jest.Mock;

  beforeAll(() => {
    // // loadRepos is mocked so that we can spy on it but also so that it doesn't trigger a network request
    // mockedLoadRepos.mockImplementation(() => ({ type: '' }));
  });

  beforeEach(() => {
    // store = configureStore({}, history);
    // mockedLoadRepos.mockClear();
  });

  afterEach(cleanup);

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = renderExchange();
    expect(firstChild).toMatchSnapshot();
  });

  it('should update input value on numpad button push', () => {
    //
  });

  it('should change incoming currency input on outgoing currency change', async () => {
    // make label for input

    // arrange
    const [refs, afterChange] = makeArrangeSet();
    let component;

    // act
    act(() => {
      component = render(
        <Exchange sliderRefs={refs} afterChange={afterChange} />,
      );
      refs[0].current!.slickNext();
    });
    await sleep(1000); // wait for interactions

    // assert
    const outgoingSlider = component.getByTestId('outgoing-slider');
    expect(outgoingSlider).toStrictEqual('?');
  });

  it('should disable an exchange button on an invalid input value (not enough assets to make a trade)', () => {
    // make a button instead of plain text
  });

  it('should update store pockets (and view) on successful exchange', () => {
    // get by test-id
  });

  // it('shouldn`t fetch repos on mount (if username is empty)', () => {
  //   renderExchange();
  //   expect(initialState.username).toBe('');
  //   expect(appActions.loadRepos).not.toHaveBeenCalled();
  // });

  // it('shouldn`t fetch repos if the form is submitted when the username is empty', () => {
  //   const { container } = renderExchange();

  //   const form = container.querySelector('form')!;
  //   fireEvent.submit(form);

  //   expect(appActions.loadRepos).not.toHaveBeenCalled();
  // });

  // it('should fetch repos if the form is submitted when the username isn`t empty', () => {
  //   const { container } = renderExchange();

  //   store.dispatch(changeUsername('julienben'));

  //   const input = container.querySelector('input')!;
  //   fireEvent.change(input, { target: { value: 'julienben' } });
  //   expect(appActions.loadRepos).not.toHaveBeenCalled();

  //   const form = container.querySelector('form')!;
  //   fireEvent.submit(form);
  //   expect(appActions.loadRepos).toHaveBeenCalled();
  // });
});
