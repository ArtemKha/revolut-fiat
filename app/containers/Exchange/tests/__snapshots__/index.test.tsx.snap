// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exchange /> should render and match the snapshot 1`] = `
<div
  class="sc-bwzfXH dFmtLp"
>
  <div
    class="sc-VigVT ervugl"
  >
    <span
      class="sc-jTzLTM cmSsxF"
    >
      <span>
        Cancel
      </span>
    </span>
    <span
      class="sc-jTzLTM cmSsxF"
    >
      <div
        class="sc-gzVnrw gftquo"
      >
        $1 = Â£
        1
        .
        21
        <span
          class="sc-htoDjs bSNund"
        >
          73
        </span>
      </div>
    </span>
    <span
      class="sc-fjdhpX clFYnx"
    >
      <span>
        Exchange
      </span>
    </span>
  </div>
  <div
    class="sc-bdVaJa hEUrkQ"
    data-testid="outgoing-slider"
  >
    <div
      class="slick-slider slick-initialized"
      dir="ltr"
    >
      <div
        class="slick-list"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="-1"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  EUR
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  102
                  .
                  05
                   
                  EUR
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  GBP
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  8
                  .
                  15
                   
                  GBP
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="2"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="3"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  EUR
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  102
                  .
                  05
                   
                  EUR
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="4"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  GBP
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  8
                  .
                  15
                   
                  GBP
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="5"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="outgoing-label"
                  for="outgoing"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  id="outgoing"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="slick-dots"
        style="display: block;"
      >
        <li
          class="slick-active"
        >
          <button>
            1
          </button>
        </li>
        <li
          class=""
        >
          <button>
            2
          </button>
        </li>
        <li
          class=""
        >
          <button>
            3
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="sc-bdVaJa cGsizg"
  >
    <div
      class="slick-slider slick-initialized"
      dir="ltr"
    >
      <div
        class="slick-list"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="-1"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="incoming-label"
                  for="incoming"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  disabled=""
                  id="incoming"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                >
                  <div
                    class="sc-gzVnrw gftquo"
                  >
                    $1 = Â£
                    0
                    .
                    89
                    <span
                      class="sc-htoDjs bSNund"
                    >
                      72
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="incoming-label"
                  for="incoming"
                >
                  GBP
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  disabled=""
                  id="incoming"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  8
                  .
                  15
                   
                  GBP
                </span>
                <span
                  class="sc-EHOje kESwbC"
                >
                  <div
                    class="sc-gzVnrw gftquo"
                  >
                    $1 = Â£
                    0
                    .
                    89
                    <span
                      class="sc-htoDjs bSNund"
                    >
                      72
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="incoming-label"
                  for="incoming"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  disabled=""
                  id="incoming"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                >
                  <div
                    class="sc-gzVnrw gftquo"
                  >
                    $1 = Â£
                    0
                    .
                    89
                    <span
                      class="sc-htoDjs bSNund"
                    >
                      72
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="2"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="incoming-label"
                  for="incoming"
                >
                  GBP
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  disabled=""
                  id="incoming"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  8
                  .
                  15
                   
                  GBP
                </span>
                <span
                  class="sc-EHOje kESwbC"
                >
                  <div
                    class="sc-gzVnrw gftquo"
                  >
                    $1 = Â£
                    0
                    .
                    89
                    <span
                      class="sc-htoDjs bSNund"
                    >
                      72
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide slick-cloned"
            data-index="3"
            style="width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <label
                  class="sc-bxivhb gxqJPj"
                  data-testid="incoming-label"
                  for="incoming"
                >
                  USD
                </label>
                <input
                  class="sc-bZQynM kwNiZT"
                  disabled=""
                  id="incoming"
                  maxlength="11"
                  size="11"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="sc-htpNat hoOIWi"
              >
                <span
                  class="sc-EHOje kESwbC"
                >
                  <span>
                    You have
                  </span>
                   
                  1023
                  .
                  35
                   
                  USD
                </span>
                <span
                  class="sc-EHOje kESwbC"
                >
                  <div
                    class="sc-gzVnrw gftquo"
                  >
                    $1 = Â£
                    0
                    .
                    89
                    <span
                      class="sc-htoDjs bSNund"
                    >
                      72
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="slick-dots"
        style="display: block;"
      >
        <li
          class="slick-active"
        >
          <button>
            1
          </button>
        </li>
        <li
          class=""
        >
          <button>
            2
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="sc-dnqmqq cOhvHF"
  >
    <div
      class="sc-gZMcBi bFxGfE"
    >
      <li
        class="sc-gqjmRU eLzNSq"
      >
        1
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        2
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        3
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        4
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        5
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        6
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        7
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        8
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        9
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
         .
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        0
      </li>
      <li
        class="sc-gqjmRU eLzNSq"
      >
        <div
          class="sc-iwsKbI jnPgxl"
        >
          <div>
            <svg
              fill="#fff"
              focusable="false"
              height="3em"
              viewBox="0 0 1024 1024"
              width="3em"
            >
              <path
                d="M689 165.1L308.2 493.5c-10.9 9.4-10.9 27.5 0 37L689 858.9c14.2 12.2 35 1.2 35-18.5V183.6c0-19.7-20.8-30.7-35-18.5z"
              />
            </svg>
          </div>
        </div>
      </li>
    </div>
  </div>
</div>
`;
